src/components/tests/SourcesTree.spec.js

```diff
@@ -85,4 +86,622 @@ describe("SourcesTree", () => {
+          currState.uncollapsedTree
+        );
+      });
+
+      it("updates tree with a new item", async () => {
+        props.sources = props.sources.merge({
+          "server1.conn13.child1/42": createMockSource(
+            "server1.conn13.child1/42",
+            "http://mdn.com/four.js"
+          )
+        });
+
+        await component.setProps({
+          ...props
+        });
+
+        expect(
+          component.state("uncollapsedTree").contents[0].contents
+        ).toHaveLength(4);
+      });
+
+      it("updates sources if sources are emptied", async () => {
+        props.sources = I.Map({});
+
+        expect(currState.uncollapsedTree.contents).toHaveLength(1);
+
+        await component.setProps({
+          ...props
+        });
+
+        expect(component.state("uncollapsedTree").contents).toHaveLength(0);
+      });
+
+      it("recreates tree if projectRoot is changed", async () => {
+        props.sources = I.Map({
+          "server1.conn13.child1/41": createMockSource(
+            "server1.conn13.child1/41",
+            "http://mozilla.com/three.js"
+          )
+        });
+        props.projectRoot = "mozilla";
+
+        expect(currState.uncollapsedTree.contents[0].contents).toHaveLength(3);
+
+        await component.setProps({
+          ...props
+        });
+
+        expect(
+          component.state("uncollapsedTree").contents[0].contents
+        ).toHaveLength(1);
+      });
+
+      it("recreates tree if debugeeUrl is changed", async () => {
+        props.sources = sourceThree;
+        props.debuggeeUrl = "mozilla";
+
+        expect(currState.uncollapsedTree.contents[0].contents).toHaveLength(3);
+
+        await component.setProps({
+          ...props
+        });
+
+        expect(
+          component.state("uncollapsedTree").contents[0].contents
+        ).toHaveLength(1);
+      });
+    });
+
+    describe("updates list items", () => {
+      beforeEach(() => {
+        props.selectLocation.mockClear();
+      });
+
+      it("updates list items if shownSource changes", async () => {
+        props.shownSource = "http://mdn.com/three.js";
+        await component.setProps({
+          ...props
+        });
+        expect(component).toMatchSnapshot();
+        expect(props.selectLocation).toHaveBeenCalledWith({
+          sourceId: "server1.conn13.child1/41"
+        });
+      });
+    });
+
+    describe("updates highlighted items", () => {
+      it("updates highlightItems if selectedSource changes", async () => {
+        props.selectedSource = sourceThree;
+        await component.setProps({
+          ...props
+        });
+        expect(component).toMatchSnapshot();
+      });
+    });
+  });
+
+  describe("focusItem", () => {
+    it("update the focused item", async () => {
+      const { component } = render({});
+
+      expect(component.state("focusedItem")).toEqual(null);
+
+      await component.instance().focusItem(itemOne);
```

Comment:
this is reaching into the internals of the component, rather than testing from the outside. a better way would be to use an event to trigger this.

Testing component internals in a unit tests makes them difficult to maintain, as these internals might change. input -> output tests are more effective and maintainable

Reply:
well, it's only used in `onKeyDown` so can look at the output of that